### PR TITLE
Add new lint for `Result<T, E>.map_or(None, Some(T))`

### LINKED ISSUE #5414: Clippy doesn't catch `Result<T,E>.map_or(None, Some(T))`
`Result<T,E>` has a method to adapt into an `Option<T>`, Clippy doesn't catch attempts to rewrite this method as `Result<T,E>.map_or(None, Some(T))`

### PR DESCRIPTION
**Pull request by nickrtorres**
Fixes #5414                                                                     
                                                                                
PR Checklist                                                                    
---                                                                             
- [x] Followed lint naming conventions (the name is a bit awkward, but it seems to conform)
- [x] Added passing UI tests (including committed .stderr file)                 
- [x] cargo test passes locally                                                 
- [x] Executed cargo dev update_lints                                           
- [x] Added lint documentation                                                  
- [x] Run cargo dev fmt                                                         
                                                                                
`Result<T, E>` has an [`ok()`](https://doc.rust-lang.org/std/result/enum.Result.html#method.ok) method that adapts a `Result<T,E>` into an `Option<T>`.
It's possible to get around this adapter by writing `Result<T,E>.map_or(None, Some)`.
                                                                                
This lint is implemented as a new variant of the existing [`option_map_none` lint](https://github.com/rust-lang/rust-clippy/pull/2128)

### CODE REVIEW

File: clippy_lints/src/methods/mod.rs [hunk 1]
```diff
@@ -330,6 +330,24 @@ declare_clippy_lint! {
     "using `Option.map_or(None, f)`, which is more succinctly expressed as `and_then(f)`"
 }
 
+declare_clippy_lint! {
+    /// **What it does:** Checks for usage of `_.map_or(None, Some)`.
+    ///
+    /// **Why is this bad?** Readability, this can be written more concisely as
+    /// `_.ok()`.
+    ///
+    /// **Known problems:**
```
Comment (flip1995) [inline clippy_lints/src/methods/mod.rs:339]: NIT:
```suggestion
    /// **Known problems:** None.
```

File: clippy_lints/src/methods/mod.rs [hunk 2]
```diff
@@ -330,6 +330,24 @@ declare_clippy_lint! {
...
+    ///
+    /// **Example:**
+    /// ```rust
+    /// # let opt = Some(1);
+    /// # let r = opt.map_or(None, Some);
```
Comment (flip1995) [inline clippy_lints/src/methods/mod.rs:344]: lines starting with `#` won't be shown. Please also add a bad and a good example

```suggestion
    /// # let opt = Some(1);
    /// let r = opt.map_or(None, Some);
```

File: clippy_lints/src/methods/mod.rs [hunk 3]
```diff
@@ -2517,37 +2536,73 @@ fn lint_map_unwrap_or_else<'a, 'tcx>(
...
+        let func_snippet = snippet(cx, map_or_args[2].span, "..");
+        let msg = "called `map_or(None, f)` on an `Option` value. This can be done more directly by calling \
+                   `and_then(f)` instead";
+        Some((
+            OPTION_MAP_OR_NONE,
+            msg,
+            "try using `and_then` instead",
+            format!("{0}.and_then({1})", self_snippet, func_snippet),
+        ))
+    } else if is_result && f_arg_is_some {
```
Comment (flip1995) [inline clippy_lints/src/methods/mod.rs:2582]:
```suggestion
    } else if is_result && default_arg_is_none && f_arg_is_some {
```

File: clippy_lints/src/methods/mod.rs [hunk 4]
```diff
@@ -2517,37 +2536,73 @@ fn lint_map_unwrap_or_else<'a, 'tcx>(
...
+
+    // This is really only needed if `is_result` holds. Computing it here
+    // makes `mess`'s assignment a bit easier, so just compute it here.
+    let f_arg_is_some = if let hir::ExprKind::Path(ref qpath) = map_or_args[2].kind {
+        match_qpath(qpath, &paths::OPTION_SOME)
+    } else {
+        false
+    };
+
+    let mess = if is_option && default_arg_is_none {
```
Comment (flip1995) [inline clippy_lints/src/methods/mod.rs:2571]: What does `mess` stand for?

I'd prefer destructuring it here and `return;` in the else case.

File: tests/ui/result_map_or_into_option.rs [hunk 5]
```diff
@@ -0,0 +1,14 @@
+// run-rustfix
+
+#![warn(clippy::result_map_or_into_option)]
+
+fn main() {
+    let opt: Result<u32, &str> = Ok(1);
+    let _ = opt.map_or(None, Some);
```
Comment (flip1995) [inline tests/ui/result_map_or_into_option.rs:7]: Please also add a test for `opt.map_or(None, |x| Some(x))`, which also should get linted and one for `opt.map_or(None, |_x| Some(y))`, which should not get linted.

Comment (nickrtorres) [reply tests/ui/result_map_or_into_option.rs:7]: Partially fixed in fb276dc3fa1d644ce6de1849ff3b78d3e54f5d9e

It looks like `opt.map_or(None, |x| Some(x))` is caught by `clippy::redundant-closure`
```
error: redundant closure found
  --> $DIR/result_map_or_into_option.rs:17:22
   |
LL |     opt.map_or(None, |x| Some(x));
   |                      ^^^^^^^^^^^ help: remove closure as shown: `Some`
   |
   = note: `-D clippy::redundant-closure` implied by `-D warnings`
```

Should I still add that test?

File: clippy_lints/src/methods/mod.rs (hunk 4 above)
Comment (nickrtorres) [reply clippy_lints/src/methods/mod.rs]: I saw `mess` in a few other lints and assumed it was a convention. I read it as "there is either `Some(mess)` that needs to be linted or no (`None`) mess that needs to be linted". 

Changed to destructure on assignment and return early in d0738bd673fe8e2b42d26b6d116f197f4aecea82

File: clippy_lints/src/methods/mod.rs (hunk 3 above)
Comment (nickrtorres) [reply clippy_lints/src/methods/mod.rs]: Fixed in d0738bd673fe8e2b42d26b6d116f197f4aecea82

File: clippy_lints/src/methods/mod.rs (hunk 2 above)
Comment (nickrtorres) [reply clippy_lints/src/methods/mod.rs]: Fixed in 3a29aedf8db3af19ee0f64dee6f00489812e6cb0

File: clippy_lints/src/methods/mod.rs (hunk 1 above)
Comment (nickrtorres) [reply clippy_lints/src/methods/mod.rs]: Fixed in 91759a7582a292c3f85204b1d61e7185c8543959

File: tests/ui/result_map_or_into_option.rs (hunk 5 above)
Comment (flip1995) [reply tests/ui/result_map_or_into_option.rs:7]: No then you don't have to lint this case. Good catch!

**Comment by flip1995** [comment]
Thanks! Waiting for GHA to get fixed.

**Pull request merged**